process/project.py

```diff
@@ -4,81 +4,183 @@
+                self.roots[oldest_child.source].append(oldest_child)
+                self.children[oldest_child.source].remove(oldest_child)
 
-    @property
-    def children(self):
-        return self.__ppts_children
+    LatestValue = namedtuple('LatestValue', ['val', 'time', 'weight'],
+                             defaults=[None, datetime.min, float('inf')])
+
+    # in priority order
+    _SOURCES = [PTS.NAME, PPTS.NAME]
 
     def field(self, name):
-        # for ppts, prefer parent record, only moving to children if none
-        # found, at which point we choose the value with the latest
-        # last_updated
+        """Fetches the value for a field, using some business logic.
+
+        The process of getting a field:
+        1. Check pts. Start with root project and descend to children. If
+           none found:
+        2. Check ppts. Start with root project and descend to children only
+           if none found on the root.
+
+        Returns:
+            string
+        """
         # TODO: I'm not even sure this is the correct logic to use for dealing
         # with ambiguities.
-        val = ''
-        if name in self.__ppts_main.values:
-            val = self.__ppts_main.values[name]['value']
-
-        update_date = datetime.min
-        if val == '':
-            for child in self.__ppts_children:
-                if name in child.values:
-                    if child.values[name]['last_updated'] > update_date:
-                        update_date = child.values[name]['last_updated']
-                        val = child.values[name]['value']
-
-        return val
+        result = self.LatestValue()
+        for (weight, source) in enumerate(self._SOURCES):
+            parents = self.roots[source]
+
+            if len(parents) > 0:
+                for parent in parents:
+                    val = parent.get_latest(name)
+                    if val and (weight < result.weight or
+                                val[1] > result.time):
+                        result = self.LatestValue(val[0], val[1], weight)
+
+                if result.val:
+                    break
+
+        latest = result
+        for (weight, source) in enumerate(self._SOURCES):
+            # if we found something on the parent, only consider higher-
+            # priority children
+            if result.val and weight >= result.weight:
+                continue
+
+            children = self.children[source]
+            if len(children) > 0:
+                for child in children:
+                    val = child.get_latest(name)
+                    if val and (weight < latest.weight or
+                                val[1] > latest.time):
+                        latest = self.LatestValue(val[0], val[1], weight)
+
+                if latest.val and (latest.weight < result.weight or
+                                   latest.time > result.time):
+                    result = latest
+                    break
```

Comment:
I think I'm just not reading this code right, but if you had to end up goign trhough multiple sets of children to get out the result, wouldn't this just break out of the first set of children and not continue on to the next source?

Reply:
this will be much simpler with the above change to require a source. the complexity of deciding between pts and ppts (and mohcd) will be handled per-field in the generators.